commit dc94645da087dc66f9e2b0bc42a69484124dbc33 (refs/changes/71/5271/2)
Author: dyep <dyep49@gmail.com>
Date:   2018-12-20T00:00:05-08:00 (10 months ago)
    
    OS-7450 VM.load should include snapshot size and free space

diff --git a/src/vm/man/vmadm.1m.md b/src/vm/man/vmadm.1m.md
index ecac3e4d..e065fe65 100644
--- a/src/vm/man/vmadm.1m.md
+++ b/src/vm/man/vmadm.1m.md
@@ -1151,6 +1151,18 @@ tab-complete UUIDs rather than having to type them out for every command.
         create: yes
         update: yes (live update)
 
+    free_space:
+
+        This specifies the amount of space in a bhyve instance that is neither
+        allocated to disks nor in use by snapshots of those disks. If snapshots
+        are present, writes to disks may reduce this value.
+
+        type: integer (number of MiB)
+        vmtype: bhyve
+        listable: no
+        create: no
+        update: no
+
     fs_allowed:
 
         This option allows you to specify filesystem types this zone is allowed
diff --git a/src/vm/node_modules/proptable.js b/src/vm/node_modules/proptable.js
index b4d191d1..df46719f 100644
--- a/src/vm/node_modules/proptable.js
+++ b/src/vm/node_modules/proptable.js
@@ -548,8 +548,8 @@ exports.properties = {
             type: 'object-array'
         },
         zfs: {
-            fields: ['compression', 'name', 'refreservation', 'volblocksize',
-                'volsize'],
+            fields: ['compression', 'name', 'refreservation', 'usedbysnapshots',
+                'volblocksize', 'volsize'],
             types: ['volume']
         },
         zonexml: 'zone.device'
@@ -855,6 +855,12 @@ exports.properties = {
             min: 0
         },
         zonexml: 'zone.attr.flexible-disk-size'
+    }, free_space: {
+        load_depends: ['flexible_disk_size'],
+        zfs: {
+            fields: ['volsize', 'usedbysnapshots'],
+            types: ['volume']
+        }
     }, fs_allowed: {
         payload: {
             allowed: {
@@ -1742,7 +1748,7 @@ exports.properties = {
         sysinfo: 'UUID'
     }, snapshots: {
         zfs: {
-            fields: ['creation', 'name'],
+            fields: ['creation', 'name', 'written'],
             types: ['snapshot']
         }
     }, set_routes: {
diff --git a/src/vm/node_modules/vmload/index.js b/src/vm/node_modules/vmload/index.js
index 87613848..0555601b 100644
--- a/src/vm/node_modules/vmload/index.js
+++ b/src/vm/node_modules/vmload/index.js
@@ -99,6 +99,10 @@ function addDatasetProperties(vmobj, dataset_objects, options, callback)
         addSnapshots(vmobj, dataset_objects, options);
     }
 
+    if (wantField(options, 'free_space') && vmobj.brand === 'bhyve') {
+        addFreeSpace(vmobj, dataset_objects, options);
+    }
+
     if (vmobj.state === 'receiving') {
         addMissingProperties(vmobj, dataset_objects, options);
     }
@@ -207,6 +211,58 @@ function addDiskProperties(vmobj, dataset_objects, options)
     });
 }
 
+/*
+ * This calculates and adds the free_space property to bhyve vmobjs
+ */
+function addFreeSpace(vmobj, dataset_objects, options)
+{
+    var diskNames;
+    var disksSize;
+    var disksSizeMiB;
+    var log = options.log;
+    var snapsSize;
+    var snapsSizeMiB;
+    var datasetsValid = true;
+
+    if (!vmobj.hasOwnProperty('flexible_disk_size')) {
+        log.trace('addFreeSpace(' + vmobj.uuid + '): '
+            + 'vmobj has no "flexible_disk_size"');
+        return;
+    }
+
+    diskNames = vmobj.disks.map(function getName(disk) {
+        return dataset_objects.mountpoints[disk.path];
+    });
+
+    for (var i = 0; i < diskNames.length; i++) {
+        var diskName = diskNames[i];
+        if (!dataset_objects.datasets[diskName]) {
+            datasetsValid = false;
+            break;
+        }
+    }
+
+    if (datasetsValid === false) {
+        log.trace('addFreeSpace(' + vmobj.uuid + '): '
+            + 'datasets do not exist');
+        return;
+    }
+
+    disksSize = diskNames.reduce(function sumVolsize(sum, name) {
+        return dataset_objects.datasets[name].volsize + sum;
+    }, 0);
+
+    snapsSize = diskNames.reduce(function sumSnaps(sum, name) {
+        return dataset_objects.datasets[name].usedbysnapshots + sum;
+    }, 0);
+
+    disksSizeMiB = disksSize / 1024 / 1024;
+    snapsSizeMiB = snapsSize / 1024 / 1024;
+
+    var freeSpace = vmobj.flexible_disk_size - disksSizeMiB - snapsSizeMiB;
+    vmobj.free_space = Number(freeSpace.toFixed(2));
+}
+
 /*
  * This populates the "missing" member of VM objects which are in state
  * "receving".
@@ -404,10 +460,12 @@ function addSnapshots(vmobj, dataset_objects, options)
     zoneroot_snapshots.forEach(function (snap) {
         var friendly_snap;
         var matches = snap.snapname.match(/^vmsnap-(.*)$/);
+        var snap_size;
         var snap_time;
 
         if (matches && matches[1]) {
-            friendly_snap = {name: matches[1]};
+            snap_size = snap.size / 1024; // in MiB
+            friendly_snap = {name: matches[1], size: snap_size};
             if (snap.hasOwnProperty('created_at')) {
                 snap_time
                     = new Date(snap.created_at * 1000); // in ms
@@ -1264,7 +1322,6 @@ function loadDatasetObjects(uuid, cache, options, callback)
     start_time = process.hrtime();
 
     getDatasets(obj, options, function (err, results) {
-
         log.debug('loading dataset_objects took %s',
             hrtime.prettyHrtime(process.hrtime(start_time)));
 
diff --git a/src/vm/node_modules/vmload/vmload-datasets.js b/src/vm/node_modules/vmload/vmload-datasets.js
index 90e5a01d..cb3ddec5 100644
--- a/src/vm/node_modules/vmload/vmload-datasets.js
+++ b/src/vm/node_modules/vmload/vmload-datasets.js
@@ -89,6 +89,9 @@ function addDatasetResult(fields, types, results, line, log)
         if (obj.hasOwnProperty('creation')) {
             snapobj.created_at = obj.creation;
         }
+        if (obj.hasOwnProperty('written')) {
+            snapobj.size = obj.written;
+        }
         if (obj.hasOwnProperty('userrefs') && obj.userrefs > 0) {
             snapobj.userrefs = obj.userrefs;
         }
@@ -328,7 +331,6 @@ function getDatasets(obj, options, callback)
 
     if (listEverything) {
         getZfsList(_options, callback);
-
     } else {
         vasync.forEachPipeline({
             inputs: parents,
@@ -526,7 +528,8 @@ function zfs(options, cmd, args, lineHandler, callback)
  *
  *     results.snapshots
  *
- *         keyed by dataset with value being array of snapname and created_at.
+ *         keyed by dataset with value being array of snapname, size, and
+ *         created_at.
  *
  *         Eg: results.snapshots['/zones/cores'] === ['snap1', ...]
  *
diff --git a/src/vm/sbin/vmadm.js b/src/vm/sbin/vmadm.js
index 4df266d5..b5c21db6 100755
--- a/src/vm/sbin/vmadm.js
+++ b/src/vm/sbin/vmadm.js
@@ -93,6 +93,8 @@ var LIST_FIELDS = {
     exit_status: {header: 'EXIT', width: 4},
     exit_timestamp: {header: 'EXIT_TIMESTAMP', width: 24},
     firewall_enabled: {header: 'FIREWALL_ENABLED', width: 16},
+    flexible_disk_size: {header: 'FLEX_DISK_SIZE', width: 14},
+    free_space: {header: 'FREE_SPACE', width: 14},
     hostname: {header: 'HOSTNAME', width: 32},
     hvm: {header: 'HVM', width: 5},
     image_uuid: {header: 'IMAGE_UUID', width: 36},
diff --git a/src/vm/tests/test-vmload-datasets.js b/src/vm/tests/test-vmload-datasets.js
index 171b59de..1ef12caf 100644
--- a/src/vm/tests/test-vmload-datasets.js
+++ b/src/vm/tests/test-vmload-datasets.js
@@ -7,7 +7,6 @@ var async = require('/usr/node/node_modules/async');
 var bunyan = require('/usr/vm/node_modules/bunyan');
 var log;
 var vmload_datasets = require('/usr/vm/node_modules/vmload/vmload-datasets');
-
 // this puts test stuff in global, so we need to tell jsl about that:
 /* jsl:import ../node_modules/nodeunit-plus/index.js */
 require('nodeunit-plus');
@@ -58,508 +57,560 @@ test('test with no datasets', function (t) {
     });
 });
 
-test('test with one zone and one image', function (t) {
-    var expected_dsobj = {
-      "datasets": {
-        "zones": {
-          "compression": "off",
-          "creation": 1521050188,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones",
-          "name": "zones",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f": {
-          "compression": "off",
-          "creation": 1521075695,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
-          "name": "zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f@final": {
-          "compression": "-",
-          "creation": 1494367539,
-          "filesystem_limit": "-",
-          "mountpoint": "/zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f@final",
-          "name": "zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f@final",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": "-",
-          "snapshot_limit": "-",
-          "type": "snapshot",
-          "userrefs": 0,
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "-"
-        },
-        "zones/2b21e6d4-af12-4966-9f42-41051949d5ba": {
-          "compression": "off",
-          "creation": 1521075876,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/2b21e6d4-af12-4966-9f42-41051949d5ba",
-          "name": "zones/2b21e6d4-af12-4966-9f42-41051949d5ba",
-          "quota": 26843545600,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/38396fc7-2472-416b-e61b-d833b32bd088": {
-          "compression": "off",
-          "creation": 1521351753,
-          "filesystem_limit": "-",
-          "mountpoint": "/dev/zvol/rdsk/zones/38396fc7-2472-416b-e61b-d833b32bd088",
-          "name": "zones/38396fc7-2472-416b-e61b-d833b32bd088",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "volume",
-          "userrefs": "-",
-          "volblocksize": 4096,
-          "volsize": 10737418240,
-          "zoned": "-"
-        },
-        "zones/38396fc7-2472-416b-e61b-d833b32bd088@final": {
-          "compression": "-",
-          "creation": 1518028564,
-          "filesystem_limit": "-",
-          "mountpoint": "/zones/38396fc7-2472-416b-e61b-d833b32bd088@final",
-          "name": "zones/38396fc7-2472-416b-e61b-d833b32bd088@final",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": "-",
-          "snapshot_limit": "-",
-          "type": "snapshot",
-          "userrefs": 0,
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "-"
-        },
-        "zones/48a81798-183f-11e8-ad9c-b7b41cb4059c": {
-          "compression": "off",
-          "creation": 1521075784,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/48a81798-183f-11e8-ad9c-b7b41cb4059c",
-          "name": "zones/48a81798-183f-11e8-ad9c-b7b41cb4059c",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/48a81798-183f-11e8-ad9c-b7b41cb4059c@final": {
-          "compression": "-",
-          "creation": 1519352017,
-          "filesystem_limit": "-",
-          "mountpoint": "/zones/48a81798-183f-11e8-ad9c-b7b41cb4059c@final",
-          "name": "zones/48a81798-183f-11e8-ad9c-b7b41cb4059c@final",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": "-",
-          "snapshot_limit": "-",
-          "type": "snapshot",
-          "userrefs": 0,
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "-"
-        },
-        "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3": {
-          "compression": "off",
-          "creation": 1521563092,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3",
-          "name": "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 10737418240,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk0": {
-          "compression": "off",
-          "creation": 1521563092,
-          "filesystem_limit": "-",
-          "mountpoint": "/dev/zvol/rdsk/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk0",
-          "name": "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk0",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": 10737418240,
-          "snapshot_limit": 18446744073709552000,
-          "type": "volume",
-          "userrefs": "-",
-          "volblocksize": 4096,
-          "volsize": 10737418240,
-          "zoned": "-"
-        },
-        "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk1": {
-          "compression": "off",
-          "creation": 1521563092,
-          "filesystem_limit": "-",
-          "mountpoint": "/dev/zvol/rdsk/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk1",
-          "name": "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk1",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": 27685552128,
-          "snapshot_limit": 18446744073709552000,
-          "type": "volume",
-          "userrefs": "-",
-          "volblocksize": 8192,
-          "volsize": 26843545600,
-          "zoned": "-"
-        },
-        "zones/archive": {
-          "compression": "lzjb",
-          "creation": 1521076431,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/archive",
-          "name": "zones/archive",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/config": {
-          "compression": "off",
-          "creation": 1521050188,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/config",
-          "name": "zones/config",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/cores": {
-          "compression": "lz4",
-          "creation": 1521050188,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "none",
-          "name": "zones/cores",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/cores/2b21e6d4-af12-4966-9f42-41051949d5ba": {
-          "compression": "lz4",
-          "creation": 1521075876,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/2b21e6d4-af12-4966-9f42-41051949d5ba/cores",
-          "name": "zones/cores/2b21e6d4-af12-4966-9f42-41051949d5ba",
-          "quota": 107374182400,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/cores/56441053-fc6d-eb86-e1b4-b94ca8e133d3": {
-          "compression": "lz4",
-          "creation": 1521563094,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/cores",
-          "name": "zones/cores/56441053-fc6d-eb86-e1b4-b94ca8e133d3",
-          "quota": 107374182400,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/cores/global": {
-          "compression": "lz4",
-          "creation": 1521050188,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/global/cores",
-          "name": "zones/cores/global",
-          "quota": 10737418240,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/dump": {
-          "compression": "off",
-          "creation": 1521050188,
-          "filesystem_limit": "-",
-          "mountpoint": "/dev/zvol/rdsk/zones/dump",
-          "name": "zones/dump",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "volume",
-          "userrefs": "-",
-          "volblocksize": 131072,
-          "volsize": 3220176896,
-          "zoned": "-"
-        },
-        "zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce": {
-          "compression": "off",
-          "creation": 1521075685,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce",
-          "name": "zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce@final": {
-          "compression": "-",
-          "creation": 1457049758,
-          "filesystem_limit": "-",
-          "mountpoint": "/zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce@final",
-          "name": "zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce@final",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": "-",
-          "snapshot_limit": "-",
-          "type": "snapshot",
-          "userrefs": 0,
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "-"
-        },
-        "zones/opt": {
-          "compression": "off",
-          "creation": 1521050188,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/opt",
-          "name": "zones/opt",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/swap": {
-          "compression": "off",
-          "creation": 1521050188,
-          "filesystem_limit": "-",
-          "mountpoint": "/dev/zvol/rdsk/zones/swap",
-          "name": "zones/swap",
-          "quota": "-",
-          "recsize": "-",
-          "refquota": "-",
-          "refreservation": 2147483648,
-          "snapshot_limit": 18446744073709552000,
-          "type": "volume",
-          "userrefs": "-",
-          "volblocksize": 8192,
-          "volsize": 6442450944,
-          "zoned": "-"
-        },
-        "zones/usbkey": {
-          "compression": "off",
-          "creation": 1521075395,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/usbkey",
-          "name": "zones/usbkey",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        },
-        "zones/var": {
-          "compression": "off",
-          "creation": 1521050188,
-          "filesystem_limit": 18446744073709552000,
-          "mountpoint": "/zones/var",
-          "name": "zones/var",
-          "quota": 0,
-          "recsize": 131072,
-          "refquota": 0,
-          "refreservation": 0,
-          "snapshot_limit": 18446744073709552000,
-          "type": "filesystem",
-          "userrefs": "-",
-          "volblocksize": "-",
-          "volsize": "-",
-          "zoned": "off"
-        }
+test('test with one zone, one bhyve VM, and four images', function (t) {
+  var expected_dsobj = {
+    "datasets": {
+      "zones": {
+        "compression": "off",
+        "creation": 1545080564,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones",
+        "name": "zones",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 540672,
+        "zoned": "off"
+      },
+      "zones/0aa46416-2c8d-658d-cb21-eba504d46270": {
+        "compression": "off",
+        "creation": 1545085213,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/0aa46416-2c8d-658d-cb21-eba504d46270",
+        "name": "zones/0aa46416-2c8d-658d-cb21-eba504d46270",
+        "quota": 5368709120,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 13321216,
+        "zoned": "off"
+      },
+      "zones/1f11188a-c71c-11e8-83d4-370e2a698b16": {
+        "compression": "off",
+        "creation": 1545080882,
+        "filesystem_limit": "-",
+        "mountpoint": "/dev/zvol/rdsk/zones/1f11188a-c71c-11e8-83d4-370e2a698b16",
+        "name": "zones/1f11188a-c71c-11e8-83d4-370e2a698b16",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "volume",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": 8192,
+        "volsize": 10737418240,
+        "written": 0,
+        "zoned": "-"
+      },
+      "zones/1f11188a-c71c-11e8-83d4-370e2a698b16@final": {
+        "compression": "-",
+        "creation": 1538578209,
+        "filesystem_limit": "-",
+        "mountpoint": "/zones/1f11188a-c71c-11e8-83d4-370e2a698b16@final",
+        "name": "zones/1f11188a-c71c-11e8-83d4-370e2a698b16@final",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": "-",
+        "snapshot_limit": "-",
+        "type": "snapshot",
+        "usedbysnapshots": "-",
+        "userrefs": 0,
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 1385337856,
+        "zoned": "-"
+      },
+      "zones/2382d24e-c75a-11e8-992b-53577424bc1a": {
+        "compression": "off",
+        "creation": 1545081043,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/2382d24e-c75a-11e8-992b-53577424bc1a",
+        "name": "zones/2382d24e-c75a-11e8-992b-53577424bc1a",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 77312,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 50176,
+        "zoned": "off"
+      },
+      "zones/2382d24e-c75a-11e8-992b-53577424bc1a@final": {
+        "compression": "-",
+        "creation": 1538605005,
+        "filesystem_limit": "-",
+        "mountpoint": "/zones/2382d24e-c75a-11e8-992b-53577424bc1a@final",
+        "name": "zones/2382d24e-c75a-11e8-992b-53577424bc1a@final",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": "-",
+        "snapshot_limit": "-",
+        "type": "snapshot",
+        "usedbysnapshots": "-",
+        "userrefs": 0,
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 98426880,
+        "zoned": "-"
+      },
+      "zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d": {
+        "compression": "off",
+        "creation": 1545081081,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d",
+        "name": "zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 0,
+        "zoned": "off"
+      },
+      "zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d@final": {
+        "compression": "-",
+        "creation": 1522855572,
+        "filesystem_limit": "-",
+        "mountpoint": "/zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d@final",
+        "name": "zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d@final",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": "-",
+        "snapshot_limit": "-",
+        "type": "snapshot",
+        "usedbysnapshots": "-",
+        "userrefs": 0,
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 441172480,
+        "zoned": "-"
+      },
+      "zones/66116621-e669-4f81-d668-bee43fd3b720": {
+        "compression": "off",
+        "creation": 1545085052,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/66116621-e669-4f81-d668-bee43fd3b720",
+        "name": "zones/66116621-e669-4f81-d668-bee43fd3b720",
+        "quota": 12705071104,
+        "recsize": 131072,
+        "refquota": 1073741824,
+        "refreservation": 1073741824,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 163840,
+        "zoned": "off"
+      },
+      "zones/66116621-e669-4f81-d668-bee43fd3b720/disk0": {
+        "compression": "off",
+        "creation": 1545085052,
+        "filesystem_limit": "-",
+        "mountpoint": "/dev/zvol/rdsk/zones/66116621-e669-4f81-d668-bee43fd3b720/disk0",
+        "name": "zones/66116621-e669-4f81-d668-bee43fd3b720/disk0",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": 11075584000,
+        "snapshot_limit": 18446744073709552000,
+        "type": "volume",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": 8192,
+        "volsize": 10737418240,
+        "written": 22599680,
+        "zoned": "-"
+      },
+      "zones/66116621-e669-4f81-d668-bee43fd3b720/disk1": {
+        "compression": "off",
+        "creation": 1545085052,
+        "filesystem_limit": "-",
+        "mountpoint": "/dev/zvol/rdsk/zones/66116621-e669-4f81-d668-bee43fd3b720/disk1",
+        "name": "zones/66116621-e669-4f81-d668-bee43fd3b720/disk1",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": 555745280,
+        "snapshot_limit": 18446744073709552000,
+        "type": "volume",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": 8192,
+        "volsize": 536870912,
+        "written": 12288,
+        "zoned": "-"
+      },
+      "zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec": {
+        "compression": "off",
+        "creation": 1545081027,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec",
+        "name": "zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 122880,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 99328,
+        "zoned": "off"
+      },
+      "zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec@final": {
+        "compression": "-",
+        "creation": 1538605331,
+        "filesystem_limit": "-",
+        "mountpoint": "/zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec@final",
+        "name": "zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec@final",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": "-",
+        "snapshot_limit": "-",
+        "type": "snapshot",
+        "usedbysnapshots": "-",
+        "userrefs": 0,
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 641669120,
+        "zoned": "-"
+      },
+      "zones/archive": {
+        "compression": "lzjb",
+        "creation": 1545080673,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/archive",
+        "name": "zones/archive",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 23552,
+        "zoned": "off"
+      },
+      "zones/config": {
+        "compression": "off",
+        "creation": 1545080570,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/config",
+        "name": "zones/config",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 48640,
+        "zoned": "off"
       },
-      "mountpoints": {
-        "/zones": "zones",
-        "/zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f": "zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
-        "/zones/2b21e6d4-af12-4966-9f42-41051949d5ba": "zones/2b21e6d4-af12-4966-9f42-41051949d5ba",
-        "/dev/zvol/rdsk/zones/38396fc7-2472-416b-e61b-d833b32bd088": "zones/38396fc7-2472-416b-e61b-d833b32bd088",
-        "/zones/48a81798-183f-11e8-ad9c-b7b41cb4059c": "zones/48a81798-183f-11e8-ad9c-b7b41cb4059c",
-        "/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3": "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3",
-        "/dev/zvol/rdsk/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk0": "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk0",
-        "/dev/zvol/rdsk/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk1": "zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk1",
-        "/zones/archive": "zones/archive",
-        "/zones/config": "zones/config",
-        "/zones/2b21e6d4-af12-4966-9f42-41051949d5ba/cores": "zones/cores/2b21e6d4-af12-4966-9f42-41051949d5ba",
-        "/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/cores": "zones/cores/56441053-fc6d-eb86-e1b4-b94ca8e133d3",
-        "/zones/global/cores": "zones/cores/global",
-        "/dev/zvol/rdsk/zones/dump": "zones/dump",
-        "/zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce": "zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce",
-        "/zones/opt": "zones/opt",
-        "/dev/zvol/rdsk/zones/swap": "zones/swap",
-        "/zones/usbkey": "zones/usbkey",
-        "/zones/var": "zones/var"
+      "zones/cores": {
+        "compression": "gzip",
+        "creation": 1545080638,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "none",
+        "name": "zones/cores",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 23552,
+        "zoned": "off"
       },
+      "zones/cores/0aa46416-2c8d-658d-cb21-eba504d46270": {
+        "compression": "gzip",
+        "creation": 1545085213,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/0aa46416-2c8d-658d-cb21-eba504d46270/cores",
+        "name": "zones/cores/0aa46416-2c8d-658d-cb21-eba504d46270",
+        "quota": 107374182400,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 23552,
+        "zoned": "off"
+      },
+      "zones/cores/66116621-e669-4f81-d668-bee43fd3b720": {
+        "compression": "gzip",
+        "creation": 1545085053,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/66116621-e669-4f81-d668-bee43fd3b720/cores",
+        "name": "zones/cores/66116621-e669-4f81-d668-bee43fd3b720",
+        "quota": 107374182400,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 23552,
+        "zoned": "off"
+      },
+      "zones/cores/global": {
+        "compression": "gzip",
+        "creation": 1545080638,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/global/cores",
+        "name": "zones/cores/global",
+        "quota": 10737418240,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 23552,
+        "zoned": "off"
+      },
+      "zones/dump": {
+        "compression": "off",
+        "creation": 1545080569,
+        "filesystem_limit": "-",
+        "mountpoint": "/dev/zvol/rdsk/zones/dump",
+        "name": "zones/dump",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "volume",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": 131072,
+        "volsize": 1073741824,
+        "written": 1073847296,
+        "zoned": "-"
+      },
+      "zones/opt": {
+        "compression": "off",
+        "creation": 1545080570,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/opt",
+        "name": "zones/opt",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 23552,
+        "zoned": "off"
+      },
+      "zones/swap": {
+        "compression": "off",
+        "creation": 1545080570,
+        "filesystem_limit": "-",
+        "mountpoint": "/dev/zvol/rdsk/zones/swap",
+        "name": "zones/swap",
+        "quota": "-",
+        "recsize": "-",
+        "refquota": "-",
+        "refreservation": 17718312960,
+        "snapshot_limit": 18446744073709552000,
+        "type": "volume",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": 8192,
+        "volsize": 17178820608,
+        "written": 12288,
+        "zoned": "-"
+      },
+      "zones/usbkey": {
+        "compression": "off",
+        "creation": 1545080570,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/usbkey",
+        "name": "zones/usbkey",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 32256,
+        "zoned": "off"
+      },
+      "zones/var": {
+        "compression": "off",
+        "creation": 1545080570,
+        "filesystem_limit": 18446744073709552000,
+        "mountpoint": "/zones/var",
+        "name": "zones/var",
+        "quota": 0,
+        "recsize": 131072,
+        "refquota": 0,
+        "refreservation": 0,
+        "snapshot_limit": 18446744073709552000,
+        "type": "filesystem",
+        "usedbysnapshots": 0,
+        "userrefs": "-",
+        "volblocksize": "-",
+        "volsize": "-",
+        "written": 1666048,
+        "zoned": "off"
+      }
+    },
+    "mountpoints": {
+     '/zones': 'zones',
+     '/zones/0aa46416-2c8d-658d-cb21-eba504d46270': 'zones/0aa46416-2c8d-658d-cb21-eba504d46270',
+     '/dev/zvol/rdsk/zones/1f11188a-c71c-11e8-83d4-370e2a698b16': 'zones/1f11188a-c71c-11e8-83d4-370e2a698b16',
+     '/zones/2382d24e-c75a-11e8-992b-53577424bc1a': 'zones/2382d24e-c75a-11e8-992b-53577424bc1a',
+     '/zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d': 'zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d',
+     '/zones/66116621-e669-4f81-d668-bee43fd3b720': 'zones/66116621-e669-4f81-d668-bee43fd3b720',
+     '/dev/zvol/rdsk/zones/66116621-e669-4f81-d668-bee43fd3b720/disk0': 'zones/66116621-e669-4f81-d668-bee43fd3b720/disk0',
+     '/dev/zvol/rdsk/zones/66116621-e669-4f81-d668-bee43fd3b720/disk1': 'zones/66116621-e669-4f81-d668-bee43fd3b720/disk1',
+     '/zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec': 'zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec',
+     '/zones/archive': 'zones/archive',
+     '/zones/config': 'zones/config',
+     '/zones/0aa46416-2c8d-658d-cb21-eba504d46270/cores': 'zones/cores/0aa46416-2c8d-658d-cb21-eba504d46270',
+     '/zones/66116621-e669-4f81-d668-bee43fd3b720/cores': 'zones/cores/66116621-e669-4f81-d668-bee43fd3b720',
+     '/zones/global/cores': 'zones/cores/global',
+     '/dev/zvol/rdsk/zones/dump': 'zones/dump',
+     '/zones/opt': 'zones/opt',
+     '/dev/zvol/rdsk/zones/swap': 'zones/swap',
+     '/zones/usbkey': 'zones/usbkey',
+     '/zones/var': 'zones/var'
+    },
       "snapshots": {
-        "zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f": [
-          {
-            "snapname": "final",
-            "dataset": "zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f",
-            "created_at": 1494367539
-          }
-        ],
-        "zones/38396fc7-2472-416b-e61b-d833b32bd088": [
-          {
-            "snapname": "final",
-            "dataset": "zones/38396fc7-2472-416b-e61b-d833b32bd088",
-            "created_at": 1518028564
-          }
-        ],
-        "zones/48a81798-183f-11e8-ad9c-b7b41cb4059c": [
-          {
-            "snapname": "final",
-            "dataset": "zones/48a81798-183f-11e8-ad9c-b7b41cb4059c",
-            "created_at": 1519352017
-          }
-        ],
-        "zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce": [
-          {
-            "snapname": "final",
-            "dataset": "zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce",
-            "created_at": 1457049758
-          }
-        ]
+      "zones/1f11188a-c71c-11e8-83d4-370e2a698b16": [
+      {
+        "snapname": "final",
+        "dataset": "zones/1f11188a-c71c-11e8-83d4-370e2a698b16",
+        "created_at": 1538578209,
+        "size": 1385337856
       }
-    };
+      ],
+      "zones/2382d24e-c75a-11e8-992b-53577424bc1a": [
+      {
+        "snapname": "final",
+        "dataset": "zones/2382d24e-c75a-11e8-992b-53577424bc1a",
+        "created_at": 1538605005,
+        "size": 98426880
+      }
+      ],
+      "zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d": [
+      {
+        "snapname": "final",
+        "dataset": "zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d",
+        "created_at": 1522855572,
+        "size": 441172480
+      }
+      ],
+      "zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec": [
+      {
+        "snapname": "final",
+        "dataset": "zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec",
+        "created_at": 1538605331,
+        "size": 641669120
+      }
+      ]
+    }
+   };
 
     /*
      * this was taken from the output of:
      *
      * zfs list -H -p -t filesystem,snapshot,volume \
      *     -o compression,creation,filesystem_limit,mountpoint,name,quota,\
-     *     recsize,refquota,refreservation,snapshot_limit,type,userrefs,volblocksize,\
-     *     volsize,zoned
+     *     recsize,refquota,refreservation,snapshot_limit,type,usedbynapshots\
+     *     userrefs,volblocksize,volsize,written,zoned
      *
      * from a system with one smartos zone, one bhyve zone and 4 installed
      * images.
      */
     var lines = [
-        'off	1521050188	18446744073709551615	/zones	zones	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'off	1521075695	18446744073709551615	/zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f	zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        '-	1494367539	-	-	zones/04a48d7d-6bb5-4e83-8c3b-e60a99e0f48f@final	-	-	-	-	-	snapshot	0	-	-	-',
-        'off	1521075876	18446744073709551615	/zones/2b21e6d4-af12-4966-9f42-41051949d5ba	zones/2b21e6d4-af12-4966-9f42-41051949d5ba	26843545600	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'off	1521351753	-	-	zones/38396fc7-2472-416b-e61b-d833b32bd088	-	-	-	0	18446744073709551615	volume	-	4096	10737418240	-',
-        '-	1518028564	-	-	zones/38396fc7-2472-416b-e61b-d833b32bd088@final	-	-	-	-	-	snapshot	0	-	-	-',
-        'off	1521075784	18446744073709551615	/zones/48a81798-183f-11e8-ad9c-b7b41cb4059c	zones/48a81798-183f-11e8-ad9c-b7b41cb4059c	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        '-	1519352017	-	-	zones/48a81798-183f-11e8-ad9c-b7b41cb4059c@final	-	-	-	-	-	snapshot	0	-	-	-',
-        'off	1521563092	18446744073709551615	/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3	zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3	0	131072	10737418240	0	18446744073709551615	filesystem	-	-	-	off',
-        'off	1521563092	-	-	zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk0	-	-	-	10737418240	18446744073709551615	volume	-	4096	10737418240	-',
-        'off	1521563092	-	-	zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/disk1	-	-	-	27685552128	18446744073709551615	volume	-	8192	26843545600	-',
-        'lzjb	1521076431	18446744073709551615	/zones/archive	zones/archive	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'off	1521050188	18446744073709551615	legacy	zones/config	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'lz4	1521050188	18446744073709551615	none	zones/cores	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'lz4	1521075876	18446744073709551615	/zones/2b21e6d4-af12-4966-9f42-41051949d5ba/cores	zones/cores/2b21e6d4-af12-4966-9f42-41051949d5ba	107374182400	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'lz4	1521563094	18446744073709551615	/zones/56441053-fc6d-eb86-e1b4-b94ca8e133d3/cores	zones/cores/56441053-fc6d-eb86-e1b4-b94ca8e133d3	107374182400	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'lz4	1521050188	18446744073709551615	/zones/global/cores	zones/cores/global	10737418240	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'off	1521050188	-	-	zones/dump	-	-	-	0	18446744073709551615	volume	-	131072	3220176896	-',
-        'off	1521075685	18446744073709551615	/zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce	zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        '-	1457049758	-	-	zones/ede31770-e19c-11e5-bb6e-3b7de3cca9ce@final	-	-	-	-	-	snapshot	0	-	-	-',
-        'off	1521050188	18446744073709551615	legacy	zones/opt	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'off	1521050188	-	-	zones/swap	-	-	-	2147483648	18446744073709551615	volume	-	8192	6442450944	-',
-        'off	1521075395	18446744073709551615	legacy	zones/usbkey	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off',
-        'off	1521050188	18446744073709551615	legacy	zones/var	0	131072	0	0	18446744073709551615	filesystem	-	-	-	off'
-    ];
+    'off     1545080564      18446744073709551615    /zones  zones   0       131072  0       0       18446744073709551615    filesystem      0       -       -       -       540672  off',
+    'off     1545085213      18446744073709551615    /zones/0aa46416-2c8d-658d-cb21-eba504d46270     zones/0aa46416-2c8d-658d-cb21-eba504d46270      5368709120      131072  0       0       18446744073709551615    filesystem      0       -       -       -       13321216        off',
+    'off     1545080882      -       -       zones/1f11188a-c71c-11e8-83d4-370e2a698b16      -       -       -       0       18446744073709551615    volume  0       -       8192    10737418240     0       -',
+    '-       1538578209      -       -       zones/1f11188a-c71c-11e8-83d4-370e2a698b16@final        -       -       -       -       -       snapshot        -       0       -       -       1385337856      -',
+    'off     1545081043      18446744073709551615    /zones/2382d24e-c75a-11e8-992b-53577424bc1a     zones/2382d24e-c75a-11e8-992b-53577424bc1a      0       131072  0       0       18446744073709551615    filesystem      77312   -       -       -       50176   off',
+    '-       1538605005      -       -       zones/2382d24e-c75a-11e8-992b-53577424bc1a@final        -       -       -       -       -       snapshot        -       0       -       -       98426880        -',
+    'off     1545081081      18446744073709551615    /zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d     zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d      0       131072  0       0       18446744073709551615    filesystem      0       -       -       -       0       off',
+    '-       1522855572      -       -       zones/63d6e664-3f1f-11e8-aef6-a3120cf8dd9d@final        -       -       -       -       -       snapshot        -       0       -       -       441172480       -',
+    'off     1545085052      18446744073709551615    /zones/66116621-e669-4f81-d668-bee43fd3b720     zones/66116621-e669-4f81-d668-bee43fd3b720      12705071104     131072  1073741824      1073741824      18446744073709551615    filesystem      0       -       -       -       163840  off',
+    'off     1545085052      -       -       zones/66116621-e669-4f81-d668-bee43fd3b720/disk0        -       -       -       11075584000     18446744073709551615    volume  0       -       8192    10737418240     22599680        -',
+    'off     1545085052      -       -       zones/66116621-e669-4f81-d668-bee43fd3b720/disk1        -       -       -       555745280       18446744073709551615    volume  0       -       8192    536870912       12288   -',
+    'off     1545081027      18446744073709551615    /zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec     zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec      0       131072  0       0       18446744073709551615    filesystem      122880  -       -       -       99328   off',
+    '-       1538605331      -       -       zones/6a449f7c-c75b-11e8-a87d-6fac326c57ec@final        -       -       -       -       -       snapshot        -       0       -       -       641669120       -',
+    'lzjb    1545080673      18446744073709551615    /zones/archive  zones/archive   0       131072  0       0       18446744073709551615    filesystem      0       -       -       -       23552   off',
+    'off     1545080570      18446744073709551615    legacy  zones/config    0       131072  0       0       18446744073709551615    filesystem      0       -       -       -       48640   off',
+    'gzip    1545080638      18446744073709551615    none    zones/cores     0       131072  0       0       18446744073709551615    filesystem      0       -       -       -       23552   off',
+    'gzip    1545085213      18446744073709551615    /zones/0aa46416-2c8d-658d-cb21-eba504d46270/cores       zones/cores/0aa46416-2c8d-658d-cb21-eba504d46270        107374182400    131072  0       0       18446744073709551615    filesystem      0       -       -       -       23552   off',
+    'gzip    1545085053      18446744073709551615    /zones/66116621-e669-4f81-d668-bee43fd3b720/cores       zones/cores/66116621-e669-4f81-d668-bee43fd3b720        107374182400    131072  0       0       18446744073709551615    filesystem      0       -       -       -       23552   off',
+    'gzip    1545080638      18446744073709551615    /zones/global/cores     zones/cores/global      10737418240     131072  0       0       18446744073709551615    filesystem      0       -       -       -       23552   off',
+    'off     1545080569      -       -       zones/dump      -       -       -       0       18446744073709551615    volume  0       -       131072  1073741824      1073847296      -',
+    'off     1545080570      18446744073709551615    legacy  zones/opt       0       131072  0       0       18446744073709551615    filesystem      0       -       -       -       23552   off',
+    'off     1545080570      -       -       zones/swap      -       -       -       17718312960     18446744073709551615    volume  0       -       8192    17178820608     12288   -',
+    'off     1545080570      18446744073709551615    legacy  zones/usbkey    0       131072  0       0       18446744073709551615    filesystem      0       -       -       -       32256   off',
+    'off     1545080570      18446744073709551615    legacy  zones/var       0       131072  0       0       18446744073709551615    filesystem      0       -       -       -       1666048 off'
+  ]
 
     var expected_args = [
         'list',
@@ -569,8 +620,8 @@ test('test with one zone and one image', function (t) {
         'filesystem,snapshot,volume',
         '-o',
         'compression,creation,filesystem_limit,mountpoint,name,quota,'
-            + 'recsize,refquota,refreservation,snapshot_limit,type,userrefs,'
-            + 'volblocksize,volsize,zoned'
+            + 'recsize,refquota,refreservation,snapshot_limit,type,'
+        + 'usedbysnapshots,userrefs,volblocksize,volsize,written,zoned'
     ];
 
     var out = {};
